File: piper/android/app/src/main/java/io/gopiper/piper/engine/statements/math_constant.kt
package io.gopiper.piper.engine.statements

import android.util.Log
import io.gopiper.piper.engine.Piper
import org.json.JSONObject
import java.lang.Exception
import kotlin.math.sqrt

suspend fun Piper.executeMathConstant(statement: JSONObject, onError: (e: Exception) -> Unit): Any? {
    try {
        Log.d("hello", "execute_math_const")
        val pointer = statement.getJSONObject("pointer")

        return when (pointer["const"].toString()) {
            "PI" -> Math.PI
            "E" -> Math.E
            "GOLDEN_RATIO" -> (1 + sqrt(5.0)) / 2
            "SQRT2" -> sqrt(2.0)
            "SQRT1_2" -> sqrt(1.0/2.0)
            "INFINITY" -> Double.POSITIVE_INFINITY
            else -> null
        }
    } catch (e: Exception) {
        onError(e)
        return null
    }
}

File: piper/extensions/chrome/src/engine/statements/math_constant.js
import { stringify } from "../util"


export default async function(statement, onError) {
    try {
        let pointer = statement.pointer
        let constant = stringify(pointer.const)

        switch(constant) {
            case "PI":
                return Math.PI
            case "E":
                return Math.E
            case "GOLDEN_RATIO":
                return (1 + Math.sqrt(5)) / 2
            case "SQRT2":
                return Math.SQRT2
            case "SQRT1_2":
                return Math.SQRT1_2
            case "INFINITY":
                return Infinity
        }
    } catch (e) {
        onError(e)
    }

    return null
}
